[DOCS-326] Split reference and how-to doc (#6468)
DataDog/integrations-core: docs/dev/new_check_howto.md

title: Create a new integration
kind: documentation
aliases:
  - /developers/integrations/integration_sdk/
  - /developers/integrations/testing/
  - /integrations/datadog_checks_dev/
  - /guides/new_integration/
---

To consider an Agent-based integration complete, and thus ready to be included in the core repository and bundled with the Agent package, a number of prerequisites must be met:

> - A `README.md` file with the right format

Reviewer: ```suggestion
- A `README.md` file with the correct format and contents
```

Let's try a dry-run using the `-n/--dry-run` flag, which won't write anything to disk.

```bash
ddev create -n awesome
```

This displays the path where the files would have been written, as well as the structure itself. For now, just make sure that the path in the _first line_ of output matches your Extras repository.

### Interactive mode

The interactive mode is a wizard for creating new integrations. By answering a handful of questions, the scaffolding will be set up and lightly pre-configured for you.

```bash
ddev create awesome
```

After answering the questions, the output matches that of the dry-run above, except in this case the scaffolding for your new integration actually exists!

## Write the check

### Intro

A Check is a Python class with the following requirements:

> - If running with Agent v7+ it should be Python 3 compatible, Python 2 otherwise for Agent v5 and v6.

Reviewer: ```suggestion
- Integrations run via Agent v7+ must be Python 3 compatible; however, Agents v5 and v6 still use Python 2.7.
```

                self.service_check('awesome.search', self.WARNING)
```

To learn more about the base Python class, see the [Python API documentation][6].

### Writing tests

There are two basic types of tests:

- Unit tests for specific functionality.
- Integration tests that execute the `check` method and verify proper metrics collection.

Tests are _required_ if you want your integration to be included in `integrations-extras`. Note that [pytest][7] and [tox][8] are used to run the tests.

For more information, see the [Datadog Checks Dev documentation][9].

#### Unit test

> The first part of the `check` method retrieves and verifies two pieces of information needed from the configuration file. This is a good candidate for a unit test. Open the file at `awesome/tests/test_awesome.py` and replace the contents with something like this:

Reviewer: ```suggestion
The first part of the `check` method retrieves and verifies two elements from the configuration file. This is a good candidate for a unit test. Open the file at `awesome/tests/test_awesome.py` and replace the contents with something like this:
```

Run only integration tests for faster development using the `-m/--marker` option:

```bash
ddev test -m integration awesome
```

The check is almost done. Let's add the final touches by adding the integration configurations.

> ### Create the check assets

Reviewer: ```suggestion
### Create the Check assets
```

> In order for your check to be complete you need to populate a set of assets provided by the ddev scaffolding . They already have the correct format but you must fill out the documents with the relevant information from the file:

Reviewer: ```suggestion
The set of assets created by the ddev scaffolding must be populated in order for a check to be considered for inclusion:
```

- **`README.md`**: This contains the documentation for your Check, how to set it up, which data it collects, etc..
- **`conf.yaml`**: This contains all configuration options for your Agent check. [See the configuration file reference documentation to learn its logic.][10]
- **`manifest.json`**: This contains the metadata for your Agent Check like its title, its categories... [See the manifest reference documentation to lean more.][11]
- **`metadata.csv`**: This contains the list of all metrics collected by your Agent Check. [See the metrics metadata reference documentation to learn more.][12]
- **`service_check.json`**: This contains the list of all Service Checks collected by your Agent check. [See the Service Check reference documentation to learn more.][13]

> For this example, those files would have the following shape:

Reviewer: ```suggestion
For this example, those files would have the following form:
```

    ## Optional flags you can set
    #
    options:
      ## @param follow_redirects - boolean - optional - default: false
      ## Set to true to follow 301 Redirect
      #
      # follow_redirects: false
```

{{% /tab %}}
{{% tab "Manifest" %}}

> The `awesome/manifest.json` for the Awesome service check. Note that the `guid` must be unique (and valid), so do _not_ use the one from this example—the tooling will generate one for you in any case:

Reviewer: ```suggestion
The `awesome/manifest.json` for the Awesome Service Check. Note that the `guid` must be unique (and valid), so do _not_ use the one from this example (the tooling will generate one for you in any case):
```
